display_list/display_list_path_effect_unittests.cc

```diff
@@ -0,0 +1,223 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/display_list_attributes_testing.h"
+#include "flutter/display_list/display_list_builder.h"
+#include "flutter/display_list/display_list_comparable.h"
+#include "flutter/display_list/types.h"
+#include "gtest/gtest.h"
+#include "include/core/SkPath.h"
+#include "include/core/SkScalar.h"
+
+namespace flutter {
+namespace testing {
+
+TEST(DisplayListPathEffect, BuilderSetGet) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  auto dash_path_effect = DlDashPathEffect::Make(TestDashes1, 2, 0.0);
+  DisplayListBuilder builder;
+  ASSERT_EQ(builder.getPathEffect(), nullptr);
+  builder.setPathEffect(dash_path_effect.get());
+  ASSERT_NE(builder.getPathEffect(), nullptr);
+  ASSERT_TRUE(Equals(builder.getPathEffect(),
+                     static_cast<DlPathEffect*>(dash_path_effect.get())));
+  builder.setPathEffect(nullptr);
+  ASSERT_EQ(builder.getPathEffect(), nullptr);
+}
+
+TEST(DisplayListPathEffect, FromSkiaNullPathEffect) {
+  std::shared_ptr<DlPathEffect> path_effect = DlPathEffect::From(nullptr);
+  ASSERT_EQ(path_effect, nullptr);
+  ASSERT_EQ(path_effect.get(), nullptr);
+}
+
+TEST(DisplayListPathEffect, FromSkiaPathEffect) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  sk_sp<SkPathEffect> sk_path_effect =
+      SkDashPathEffect::Make(TestDashes2, 2, 0.0);
+  std::shared_ptr<DlPathEffect> dl_path_effect =
+      DlPathEffect::From(sk_path_effect);
+
+  ASSERT_EQ(dl_path_effect->type(), DlPathEffectType::kDash);
+  ASSERT_TRUE(
+      Equals(dl_path_effect, DlDashPathEffect::Make(TestDashes2, 2, 0.0)));
+}
+
+TEST(DisplayListPathEffect, EffectShared) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_TRUE(Equals(effect->shared(), effect));
+}
+
+TEST(DisplayListPathEffect, DashEffectAsDash) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_NE(effect->asDash(), nullptr);
+  ASSERT_EQ(effect->asDash(), effect.get());
+}
+
+TEST(DisplayListPathEffect, DashEffectEquals) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect1 = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  auto effect2 = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_TRUE(Equals(effect1, effect2));
```

Comment:
In the FooEquals, and FooNotEquals tests (the others can just use ASSERT_EQ):

Use TestEquals here (and TestNotEquals in the not-equal tests) as defined in "display_list_attributes_testing.h" as they will test a lot of equals conditions, including:

- == bidirectionally
- != bidirectionally
- ASSERT_EQ works bidirectionally
- Equals bidirectionally

Also beware that both "ASSERT_EQ" and "TestEquals" will be doing pointer compares if you hand them a pointer - and those won't do a deep inspection on the object being pointed to. The shared_ptr's used here also act as pointers that way. So, be sure to dereference the arguments when you want to compare the objects.

The few places where you don't want to dereference is when you actually want to compare some pointers (usually to nullptr) such as when you are testing whether "asDash" returned a pointer or a null.

Reply:
yes, the Equal method will receiving `shared_ptr` and future will call the `DlDashPathEffect::equals_()`